### ISSUE TITLE
The current CUDA stream is not honored by some operations

### ISSUE THREAD

**Issue by yongjik** [pytorch/pytorch#2702]
Hi, I think something is wrong with handling CUDA streams.  Could someone take a look?

(I'm using Python 2.7.12, CUDA 8.0, and PyTorch version 0.2.0.post1)

    #!/usr/bin/python
    # PyTorch CUDA stream test.

    import torch

    DIM = 2048
    M = torch.cuda.FloatTensor(DIM, DIM)

    def do_test(M, use_stream, fn):
        M.fill_(0.1)
        print 'Initial value = %.4f' % fn()

        # Create a new stream if requested.
        stream = torch.cuda.Stream() if use_stream else None
        with torch.cuda.stream(stream):
            for step in range(5):
                # Do random calculation.
                K = torch.matmul(M, M) / DIM
                M -= K
                print 'After step %d = %.4f' % (step, fn())

    fn = M.norm   # or 'M.sum' or 'lambda: M[0, 0]'

    print '=== Without using a new stream ==='
    do_test(M, False, fn)
    print '=== With using a new stream ==='
    do_test(M, True, fn)

Result on my machine (GTX 1080).  The second half keeps changing:

    === Without using a new stream ===
    Initial value = 204.8000
    After step 0 = 184.3197
    After step 1 = 167.7312
    After step 2 = 153.9943
    After step 3 = 142.4149
    After step 4 = 132.5118
    === With using a new stream ===
    Initial value = 204.8000
    After step 0 = 3.1469
    After step 1 = 3.1469
    After step 2 = 3.1469
    After step 3 = 3.1469
    After step 4 = 3.1469

Basically, when I'm using a different stream, computing any single float value returns random results.  Running nvprof shows that when the second do_test() runs, all the calculations happen in the new stream (13 below), but copying the result back to CPU happens in the default stream (7 below), and at the wrong moments before the correct value is ready.

    $ nvprof --print-gpu-trace ./cuda-stream-test.py

       Start  Duration  Size  Stream  Name
    ... (only showing the second half)
    2.28038s     320ns    4B       7  [CUDA memcpy DtoH]
    2.28086s  2.0956ms     -      13  sgemm_128x128x8_NN_vec [796]
    2.28132s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28183s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28194s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28206s  1.0560us    4B       7  [CUDA memcpy DtoH]
    2.28218s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28296s  147.40us     -      13  void kernelPointwiseApply2<TensorDivConstantOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorDivConstantOp<float>, float>, TensorInfo<float, float>, float, float) [814]
    2.28311s  212.52us     -      13  void kernelPointwiseApply2<TensorSubOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorSubOp<float>, float>, TensorInfo<float, float>, float, float) [827]
    2.28332s  93.890us     -      13  void kernelReduceAllPass1<TensorNormOp<float, int=2>, ReduceAdd<float, float>, ReduceAdd<float, float>, float, float, unsigned int, int=-2>(TensorInfo<float, ReduceAdd<float, float>>, ReduceAdd<float, float>, float, float, int=2, TensorNormOp<float, int=2>, float*) [857]
    2.28341s  2.8160us     -      13  void kernelReduceAllPass2<ReduceAdd<float, float>, float, unsigned int>(int, float, float, float*, float*) [866]
    2.28342s  2.0951ms     -      13  sgemm_128x128x8_NN_vec [892]
    ... (4 more calls to these kernels)

Please enlighten me if I'm doing something wrong.  Thanks!

**Comment by zou3519** [comment]
@yongjik I'm looking into this: there definitely seems to be a race condition somewhere.

In the meantime, you could add `stream.synchronize()` before the print statement and I think the code will behave as you wanted it to.

**Comment by ngimel** [comment]
This is likely because new streams are created with CudaStreamNonBlocking https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/Stream.cpp#L18, and as @yongjik noted D2H copy happens in the default stream that does not synchronize with the non-blocking stream. Looks like a bug.

**Comment by yongjik** [comment]
Hi @zou3519, thanks for looking into this.

FYI, I looked at the code a bit, and I think there are several places where the code calls `cudaMemcpy()`, which always executes on the default stream, even when we might have a non-default stream.  I played around a bit and the following seems to work:

```diff
diff --git a/torch/lib/THC/THCReduceAll.cuh b/torch/lib/THC/THCReduceAll.cuh
index a6e1e85..5c57c24 100644
--- a/torch/lib/THC/THCReduceAll.cuh
+++ b/torch/lib/THC/THCReduceAll.cuh
@@ -331,7 +331,10 @@ bool THC_reduceAll(THCState* state,
   // If our destination is not on the device, copy the value back to
   // the host (synchronous!)
   if (!outOnDevice) {
-    THCudaCheck(cudaMemcpy(out, devOut, sizeof(AccT), cudaMemcpyDeviceToHost));
+    cudaStream_t stream = THCState_getCurrentStream(state);
+    THCudaCheck(cudaMemcpyAsync(out, devOut, sizeof(AccT),
+                                cudaMemcpyDeviceToHost, stream));
+    THCudaCheck(cudaStreamSynchronize(stream));
   }
 
   if (freeDevOut) {
diff --git a/torch/lib/THC/generic/THCStorage.c b/torch/lib/THC/generic/THCStorage.c
index 8a6cede..beb072a 100644
--- a/torch/lib/THC/generic/THCStorage.c
+++ b/torch/lib/THC/generic/THCStorage.c
@@ -28,8 +28,12 @@ real THCStorage_(get)(THCState *state, const THCStorage *self, ptrdiff_t index)
 {
   THArgCheck((index >= 0) && (index < self->size), 2, "index out of bounds");
   real value;
-  THCudaCheck(cudaMemcpy(&value, self->data + index, sizeof(real),
-                         cudaMemcpyDeviceToHost));
+
+  cudaStream_t stream = THCState_getCurrentStream(state);
+  THCudaCheck(cudaMemcpyAsync(&value, self->data + index, sizeof(real),
+                              cudaMemcpyDeviceToHost, stream));
+  THCudaCheck(cudaStreamSynchronize(stream));
+
   return value;
 }
 
diff --git a/torch/lib/THC/generic/THCTensorCopy.c b/torch/lib/THC/generic/THCTensorCopy.c
index b5122b8..b23e013 100644
--- a/torch/lib/THC/generic/THCTensorCopy.c
+++ b/torch/lib/THC/generic/THCTensorCopy.c
@@ -59,10 +59,13 @@ void THTensor_(copyCuda)(THCState *state, THTensor *self, struct THCTensor *src)
     THTensor *selfc = THTensor_(newContiguous)(self);
     src = THCTensor_(newContiguous)(state, src);
 
-    THCudaCheck(cudaMemcpy(THTensor_(data)(selfc),
-                           THCTensor_(data)(state, src),
-                           THCTensor_(nElement)(state, src) * sizeof(real),
-                           cudaMemcpyDeviceToHost));
+    cudaStream_t stream = THCState_getCurrentStream(state);
+    THCudaCheck(cudaMemcpyAsync(THTensor_(data)(selfc),
+                                THCTensor_(data)(state, src),
+                                THCTensor_(nElement)(state, src) * sizeof(real),
+                                cudaMemcpyDeviceToHost,
+                                stream));
+    THCudaCheck(cudaStreamSynchronize(stream));
 
     THCTensor_(free)(state, src);
     THTensor_(freeCopyTo)(selfc, self);
```

Each fix applies to: reduce operations (such as `sum()`), picking up an element (`A[0, 0]`), and `copy_()`.  However, there are more places in the code that uses `cudaMemcpy()` and I'm not sure I understand the rest of them.

**Comment by zou3519** [comment]
@ngimel thanks for identifying the bug!

@yongjik yes, that fixes the bug (and probably is an efficiency win).
I think for now though I'd rather keep the DtoH transfers on the default stream (and think about moving DtoH transfers to the current stream later for efficiency gains). Creating new streams with cudaStreamNonBlocking set by default seems like it could lead to a bunch of problems so I think turning that off is the way to go

**Comment by colesbury** [comment]
@zou3519, we should make sure that the copies happen on the current stream

**Comment by zou3519** [comment]
Sounds good, thanks!

**Comment by andrei-pokrovsky** [comment]
Another possibility to keep in mind is one of custom ops is not respecting your default stream - every custom op should be properly implemented to respect pytorch stream.
